### ISSUE TITLE
Flaky E2E: Login sometimes hangs on network idle when running E2E tests locally

### ISSUE THREAD

**Issue by dpasque** [Automattic/wp-calypso#69294]
### Spec file

Happens in login-page.ts

### Logs

Context: p1665671451608789-slack-CQD1HH4MA

It seems like inconsistently for some people, running E2E tests locally causes the tests to timeout when waiting for the login page to fire the `networkidle` event state. 

We should figure out why this might be happening for some folks, and how we can get around it. Do we even need the `networkidle` wait anymore?

**Comment by worldomonation** [comment]
I've reassigned the project from "General Maintenance" to "E2E Flakiness".

**Comment by karenroldan** [comment]
`networkidle` is consistently failing for me when running the tests locally.  

> consider operation to be finished when there are no network connections for at least 500 ms. ([Source](https://playwright.dev/docs/api/class-page#page-goto))

### What I found 
On the `/log-in` page, I am getting a **pending** status for `https://wordpress.com/cspreport`

```
Request headers:
Content-Type: application/csp-report
Referer: https://wordpress.com/
sec-ch-ua: "Google Chrome";v="107", "Chromium";v="107", "Not=A?Brand";v="24"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "macOS"
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36
```
this means I still have open network connections, thus the `networkidle` causes the tests to timeout.

I don't get the timeout with `networkidle` when I add the route. This is not the solution that we need, but this confirms the open network that we have.

```
async visit( { path }: { path: string } = { path: '' } ): Promise< Response | null > {
	const targetUrl = path ? `log-in/${ path }` : 'log-in';
	this.page.route( '**/csreport', ( route ) => route.continue() );
	return await this.page.goto( getCalypsoURL( targetUrl ), { waitUntil: 'networkidle' } );
} 
```

### Why this might be happening for some folks

I couldn't figure out exactly why it is happening for some people and not all. 

### How we can get around it? Do we even need the `networkidle` wait anymore?

`domcontentloaded` and `load` events could fire very quickly.
But by default, `page.goto` will wait for the `load` event. 
`networkidle` is a more heuristic approach; it will wait for there to be **no network activity** for at least 500ms. But, this is flaky for us. 

**Option 1:** We can use `waitUntil: 'commit'` where the `page.goto` operation is considered to be finished when the network response is received and the document started loading.

**Option 2:** Create an implementation that will manually wait for pending requests. 

**Option 3:** Add a logic to only track requests without `Content-Type: application/csp-report`

**Comment by dpasque** [comment]
@karenroldan Awesome investigation! 🎉  thank you so much for the detailed dive and the notes!

TL;DR -- I'm not too worried about the CSP report call, and I think your approach of actually just overriding that request might not be a bad way forward! I think we could also just wait for `load` event if that ends up being enough!

> On the /log-in page, I am getting a pending status for https://wordpress.com/cspreport

This is so bizarre to me! I wonder if there's some bug here in our Calypso web server? [Looking at the route handler here](https://github.com/Automattic/wp-calypso/blob/trunk/client/server/pages/index.js#L894), it really doesn't do much at all, and doesn't do anything that would take time, so not sure why it's hanging. 🤔 

I looked at the Tracks event history for the report, and we've had no drastic changes in the volume of these were logging (and there are literally several million a day!). I don't see any owners of the event in Tracks, so I don't know exactly what is expected here. Maybe that's more than we expect and we're slogging down the server? @noahtallen -- do you know who looks at the CSP violations?

> Option 1: We can use waitUntil: 'commit' where the page.goto operation is considered to be finished when the network response is received and the document started loading.

I could be misunderstanding, but I think this will fire even before `load`? That's based on their [descriptions here](https://playwright.dev/docs/navigations).

`load` might be enough here though. How was that working in your local tests? @worldomonation, do you remember why we originally opted to wait for `networkidle` here?

> Add a logic to only track requests without Content-Type: application/csp-report

I think this is a good idea! And honestly, overriding the response like you did with `page.route` seems like a totally valid approach to me!

**Comment by worldomonation** [comment]
> `load` might be enough here though. How was that working in your local tests? @worldomonation, do you remember why we originally opted to wait for `networkidle` here?

I can tell you about it, certainly.

Back when the framework was being developed we kept running into issues once the login process moved on from the `/log-in` endpoint, through to the intermediate loading page and then finally to `/home`. This was all on **simple** sites.

Not certain whether this is still the case for Atomic vs Simple, but it used to be that on Simple sites the `networkidle` was one of the ways to wait for everything to be loaded in Calypso (including lazy-loaded items such as offers in the sidebar). Waiting only for the more common (and less aggressive) `load` status often led to Playwright attempting to perform actions on the page before everything had loaded, causing intermittent failures.

Now that we're using more direct navigation to the starting point on top of using saved cookie authentication to bypass the login redirects in the first place, I think we can try moving to a `load` wait in conjunction with the network call intercept that @karenroldan has proposed and it _should not_ have much of an impact.

**Comment by noahtallen** [comment]
> Maybe that's more than we expect and we're slogging down the server? @noahtallen -- do you know who looks at the CSP violations?

No clue, unfortunately. 

> Maybe that's more than we expect and we're slogging down the server

I am seeing a huge number of cspreport requests on the log-in page, but I guess those are sent by the browser and not by us directly. It does seem like a good idea to fix them where possible.

However, I also looked into our logs, and the route only timed out twice in the past 7 days. In the same logs, there appear to have been 91.6 million requests to /cspreport, so I don't know if this would show up enough to impact tests. However, I don't think these logs cover calypso.live, just production and staging.

In other words, if we were bogging down the server, it seems weird that it wouldn't show up in production.

> This is so bizarre to me! I wonder if there's some bug here in our Calypso web server?

It definitely seems possible, but this is such a straightforward route it's weird! I wonder if there is something weird about how the browser in the test suite is handling this? I couldn't reproduce myself, unfortunately :/

**Comment by noahtallen** [comment]
A possible solution is to basically do nothing if the environment isn't production or staging. For example, we already don't log to tracks in dev mode. Perhaps we could just return immediately at the top of the handler if we're on wpcalypso or similar.

**Comment by dpasque** [comment]
> In other words, if we were bogging down the server, it seems weird that it wouldn't show up in production.

Yeah, very much agreed!

> I wonder if there is something weird about how the browser in the test suite is handling this?

This is what I was wondering too. Either way, it doesn't seem like there's any meaningful disruption on the application side, so I'm thinking our best bet is just tailoring our test script to work around any hangups.

**Comment by karenroldan** [comment]
> I could be misunderstanding, but I think this will fire even before load? That's based on their [descriptions here](https://playwright.dev/docs/navigations).

Yes, you are right! Thanks for the info. The `commit` event might cause more flaky tests if that's the case. We could try out the default event. 

> load might be enough here though. How was that working in your local tests?

I didn't get any failures with the flow from the login when using `load`. Similar to `domcontentloaded` and `commit`. 

> Now that we're using more direct navigation to the starting point on top of using saved cookie authentication to bypass the login redirects in the first place, I think we can try moving to a load wait in conjunction with the network call intercept

Good to know! I first created a draft with the changes that use the `load` wait and network call intercept. I'd check and see first if it did not affect the tests. 

> This is what I was wondering too. Either way, it doesn't seem like there's any meaningful disruption on the application side, so I'm thinking our best bet is just tailoring our test script to work around any hangups.

💯 using a workaround would be a good start. 

Thanks for your input!

**Comment by karenroldan** [comment]
Done in #70149.
